File: backend/backend/src/main/java/com/sareekart/service/ProductService.java
package com.sareekart.service;

import com.sareekart.dto.request.ProductRequest;
import com.sareekart.dto.request.ProductSearchCriteria;
import com.sareekart.dto.response.PagedResponse;
import com.sareekart.dto.response.ProductResponse;
import java.math.BigDecimal;

public interface ProductService {

    PagedResponse<ProductResponse> getAllProducts(int page, int size, String sortBy, String sortDir);

    ProductResponse getProductById(Long id);

    ProductResponse createProduct(ProductRequest request);

    ProductResponse updateProduct(Long id, ProductRequest request);

    void deleteProduct(Long id);

    PagedResponse<ProductResponse> searchProducts(String query, int page, int size, String sortBy, String sortDir);

    PagedResponse<ProductResponse> filterProducts(Long categoryId, BigDecimal minPrice, BigDecimal maxPrice, String fabric, int page, int size, String sortBy, String sortDir);

    PagedResponse<ProductResponse> searchAndFilterProducts(ProductSearchCriteria criteria);
}

File: frontend/frontend/src/services/productService.js
import api from '../api/axiosConfig';

/**
 * Product API service for SareeKart
 */
const productService = {
  /**
   * Get all products with pagination and sorting
   */
  getProducts: async (params = {}) => {
    const { page = 0, size = 12, sortBy = 'createdAt', sortDir = 'desc' } = params;
    const response = await api.get('/products', {
      params: { page, size, sortBy, sortDir },
    });
    return response.data;
  },

  /**
   * Get a single product by ID
   */
  getProductById: async (id) => {
    const response = await api.get(`/products/${id}`);
    return response.data;
  },

  /**
   * Search products by query string
   */
  searchProducts: async (query, params = {}) => {
    const { page = 0, size = 12 } = params;
    const response = await api.get('/products/search', {
      params: { q: query, page, size },
    });
    return response.data;
  },

  /**
   * Filter products by category, price, fabric
   */
  filterProducts: async (filters = {}) => {
    const { categoryId, minPrice, maxPrice, fabric, page = 0, size = 12, sortBy = 'createdAt', sortDir = 'desc' } = filters;
    const response = await api.get('/products/filter', {
      params: { categoryId, minPrice, maxPrice, fabric, page, size, sortBy, sortDir },
    });
    return response.data;
  },

  /**
   * Create a new product (Admin)
   */
  createProduct: async (productData) => {
    const response = await api.post('/admin/products', productData);
    return response.data;
  },

  /**
   * Update an existing product (Admin)
   */
  updateProduct: async (id, productData) => {
    const response = await api.put(`/admin/products/${id}`, productData);
    return response.data;
  },

  /**
   * Delete a product (Admin)
   */
  deleteProduct: async (id) => {
    const response = await api.delete(`/admin/products/${id}`);
    return response.data;
  },
};

export default productService;
